pkg/controller/statefulset/stateful_set.go

```diff
@@ -163,15 +180,41 @@ func (ssc *StatefulSetController) updatePod(old, cur interface{}) {
 		// Two different versions of the same pod will always have different RVs.
 		return
 	}
-	set := ssc.getStatefulSetForPod(curPod)
-	if set == nil {
+	glog.V(4).Infof("Pod %s updated, objectMeta %+v -> %+v.", curPod.Name, oldPod.ObjectMeta, curPod.ObjectMeta)
+
+	labelChanged := !reflect.DeepEqual(curPod.Labels, oldPod.Labels)
+
+	curControllerRef := controller.GetControllerOf(curPod)
+	oldControllerRef := controller.GetControllerOf(oldPod)
+	controllerRefChanged := !reflect.DeepEqual(curControllerRef, oldControllerRef)
+	if controllerRefChanged {
+		// The ControllerRef was changed. Sync the old controller, if any.
+		if oldControllerRef != nil {
+			set, err := ssc.setLister.StatefulSets(oldPod.Namespace).Get(oldControllerRef.Name)
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf("Couldn't get StatefulSet %q in namespace %q: %v", oldControllerRef.Name, oldPod.Namespace, err))
+			} else {
+				ssc.enqueueStatefulSet(set)
+			}
+		}
+	}
+
+	// If it has a ControllerRef, that's all that matters.
+	if curControllerRef != nil {
+		set, err := ssc.setLister.StatefulSets(curPod.Namespace).Get(curControllerRef.Name)
+		if err != nil {
```

Comment:
Technically you should be handling NotFound here, but I don't think there's a downside.